src/libraries/Native/Unix/System.Native/pal_io.c

```diff
@@ -1403,6 +1403,10 @@ static bool CopyFile_CopyFileRange(int inFd, int outFd, const struct stat_ *sour
             // so return false to try a legacy read(2)/write(2) copy.
             return false;
         }
+        else if (copied == 0)
+        {
+            break;
```

Comment:
Should this case be handled like this, tough? If the number of bytes copied is 0, but there's still stuff to copy, it's safe to let the loop finish from the `while (size)` condition. Otherwise, we risk a file being silently truncated if `size != 0` and `copy_file_range()` returns `0` for whatever reason.

Reply:
The documentation for `copy_file_range` seems to suggest that zero is returned only for the end of input file. I'd suggest to not depend on the `size` at all (ie. also remove it from the `while` loop) and treat it much like the read/write fallback loop.

Reply:
If a file is truncated while being copied (which is possible because the locks are just advisory), then, yes, this is a good way of handling the input EOF.